### PR TITLE
feat-#429: Added Signin with Google button

### LINKED ISSUE #429: [FEATURE] :Want to Add signin with google button
### Description

want to add signin with google button

### Use Case

it will make login /signup easier

### Proposed Solution

_No response_

### Additional Information

_No response_

### Attachments

_No response_

### PR DESCRIPTION
**Pull request by siddharthisrani**
Summary
This PR introduces the 'Sign in with Google' button to the login page, enabling users to authenticate using their Google accounts.

Description
This PR addresses the need for Google authentication within the WanderLust application. The changes include:

   Addition of 'Sign in with Google' Button:
    Implemented a new button on the login page to initiate Google sign-in.

   Google OAuth2 Strategy Integration:
    Integrated Passport.js with the Google OAuth2 strategy for handling authentication.

  Google OAuth2 Configuration:  
    Configured the application to use Google OAuth2 credentials (client ID and secret) via environment variables.

  OAuth2 Callback Handling:
     Implemented the logic to handle the Google OAuth2 callback, authenticate users, and create user sessions.

  Frontend UI Update:
     Updated the frontend to display 'Create Post' and 'Logout' buttons upon successful Google login.

Technical details:
   The Google OAuth2 client ID and secret are securely managed through environment variables.
   Passport.js is used to manage the Google authentication strategy.
   The frontend dynamically updates to reflect the user's authenticated state.

Issue(s) Addressed
Closes #429

Prerequisites
 Yes followed all the [CONTRIBUTING GUIDELINES](https://github.com/krishnaacharyaa/wanderlust/blob/main/.github/CONTRIBUTING.md#guidelines-for-contributions)?

### CODE REVIEW

**Comment by krishnaacharyaa** [comment]
Please attach a video of a working demo

**Comment by krishnaacharyaa** [comment]
@Sukomal07 can you please review

**Comment by krishnaacharyaa** [comment]
@Deblin798 feel free to review

**Comment by siddharthisrani** [comment]
> Please attach a video of a working demo


https://github.com/user-attachments/assets/c703c1e3-7723-4f36-90c8-b481ddb8f438

File: backend/config/passport.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import { Strategy as GoogleStrategy } from 'passport-google-oauth20';
+import User from '../models/user.js';
+
+passport.use(
+  new GoogleStrategy(
+    {
+      clientID: process.env.GOOGLE_CLIENT_ID,
+      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
+      callbackURL: 'http://localhost:8080/api/auth/google/callback',
+    },
```
Comment (krishnaacharyaa) [inline backend/config/passport.js:11]: As we are deploying in vercel, we can't hardcode this to 8080, kindly move this to evn variable or in this case i guess you need to use the backend_url env variable

File: backend/routes/auth.js [hunk 2]
```diff
@@ -13,10 +15,38 @@ const router = Router();
...
+  passport.authenticate('google', { failureRedirect: '/' }),
+  (req, res) => {
+    const token = jwt.sign({ id: req.user._id }, process.env.JWT_SECRET, { expiresIn: '1h' });
+    res.cookie('access_token', token, {
+      httpOnly: true,
+      secure: process.env.NODE_ENV === 'production',
+      sameSite: 'lax',
+    });
+    res.redirect('http://localhost:5173/signup?google-callback=true');
+  }
```
Comment (krishnaacharyaa) [inline backend/routes/auth.js:32]: Same here, kindly use frontend url env variable

File: frontend/src/pages/signup-page.tsx [hunk 3]
```diff
@@ -63,11 +64,49 @@ function Signup() {
...
+        }
+      }
+    };
+
+    handleGoogleCallback();
+  }, [location, navigate]);
+
+  const handleGoogleLogin = () => {
+    window.location.href = 'http://localhost:8080/api/auth/google';
+  };
```
Comment (krishnaacharyaa) [inline frontend/src/pages/signup-page.tsx:103]: Same here

File: frontend/src/pages/signup-page.tsx [hunk 4]
```diff
@@ -63,11 +64,49 @@ function Signup() {
...
+    const handleGoogleCallback = async () => {
+      const searchParams = new URLSearchParams(location.search);
+      const isGoogleCallback = searchParams.get('google-callback') === 'true';
+
+      if (isGoogleCallback && !toastShownRef.current) {
+        try {
+          const response = await axiosInstance.get('/api/auth/check');
+          const { token, user } = response.data;
+          localStorage.setItem('token', token);
+          if (user && user._id && user.role) {
```
Comment (krishnaacharyaa) [inline frontend/src/pages/signup-page.tsx:77]: localstorage we are setting token not in the cookies? @Sukomal07 kindly confirm

**Review by krishnaacharyaa — CHANGES_REQUESTED** [review]
Thank you for the PR @siddharthisrani 
Just couple of review comments, kindly address
And if any env variables are used, add it in the .env.sample

**Comment by siddharthisrani** [comment]
Hii @krishnaacharyaa the requested changes to use environment variables for URLs in Google OAuth and redirects 

Changes Made:
Used environment variables for URLs in Google OAuth and redirects.
Updated .env.sample to include the new environment variables.

**Comment by krishnaacharyaa** [comment]
@Sukomal07 can you please have a look at this PR

**Comment by Sukomal07** [comment]
> @Sukomal07 can you please have a look at this PR

@siddharthisrani didn't reply to my question.

**Comment by krishnaacharyaa** [comment]
@siddharthisrani please sync up with @Sukomal07 , so that we can merge this

**Comment by siddharthisrani** [comment]
> @siddharthisrani didn't reply to my question.
Hi @Sukomal07  and @krishnaacharyaa ,

Regarding the question about token storage:

In the current implementation, the token is being set in local storage after a successful Google OAuth login. This was done to maintain consistency with other parts of the application that also use local storage for token management.

However, from the backend, we are also sending the token in cookies to enhance security and manage user sessions more effectively. Here’s the relevant code:

router.get(
  '/google/callback',
  passport.authenticate('google', { failureRedirect: '/' }),
  (req, res) => {
    const token = jwt.sign({ id: req.user._id }, process.env.JWT_SECRET, { expiresIn: '1h' });
    res.cookie('access_token', token, {
      httpOnly: true,
      
Cookie

This implementation ensures that the token is available in cookies for enhanced security, while local storage is used to maintain consistency in token management across the application.

Please let me know if my answer is according to your requirements. If any further changes needed i will be happy to assist you.

**Comment by siddharthisrani** [comment]
> @siddharthisrani can we do without saving token in localstorage ?

Sure ,i will look in this part for solutions and get back to you .

**Comment by siddharthisrani** [comment]
Hii @Sukomal07 
Yes, it's working well without saving the token in local storage.
localStorage
removedLocalStorage

**Comment by Sukomal07** [comment]
@siddharthisrani Nice , then remove localstorage part from all of your code.

**Comment by siddharthisrani** [comment]
Hii @Sukomal07,

I've removed the token saving in local storage from all the relevant parts of the code. Please review the latest commits.

**Review by Sukomal07 — APPROVED** [review]

**Review by Sukomal07 — APPROVED** [review]

**Review by krishnaacharyaa — APPROVED** [review]
Thank you @siddharthisrani for your active contribution

**Pull request merged**